### PR TITLE
Refactor how Rails handles filter categories passed to and from Vue

### PR DESCRIPTION
**Pull request by h-m-m**
### Why
The contribution filtering works nicely, but difficult to understand. We'd like to add new filters, and that's hard at the moment. I wrote the original code and am unhappy with how confusing it is. Its confusing complexity is what makes adding or editing filters more difficult for us, and also slows down bringing people into the project and up to speed on how all this stuff works. 

### What

The big change is making each type of filter its own class.
Each class
- accepts an ActiveRecord relation
- knows how to filter that relation by request parameters passed from the controller

Also used this opportunity to:
- make the controller code easier to follow
- simplify the logic around conditionally passing links to the frontend (do we show the link to contact someone? does the user have permission to see this?
- work on our nomenclature for clarity & simplicity
 
And hopefully this will make adding new filters easier in the future

### Testing

Updated some tests. Could probably use more work in this regard

### Next Steps

Some next steps are already called out in the comments when this PR was in Draft status.
Also: add a filter for Date now that it should be easier to do so

### Accessibility

We were very careful to leave the existing UI unchanged in any visible way, so this should be accessibility-neutral

### Security

Hopefully this improves the security what with less dynamic/functional code and code that's easier for future devs to maintain

### Pre-Merge Checklist

- [x] Security & accessibility have been considered
- [ ] Tests have been added, or an explanation has been given why the features cannot be tested
- [x] Documentation and comments have been added to the codebase where required
- [x] Entry added to CHANGELOG.md if appropriate
- [x] Outstanding questions and concerns have been resolved
- [ ] Any next steps have been turned into Issues or Discussions as appropriate

### CODE REVIEW

File: app/filters/contribution_type_filter.rb
Comment (h-m-m) [inline app/filters/contribution_type_filter.rb:19]: Is this right — that `matchable` is what we want here?

**Review by h-m-m — COMMENTED** [review]
I added a bunch of comments to stuff because it helps me.

Sorry, @maebeale, for the flipping back and forth about what goes in instances and what goes in classes. I added on to your work with what I thought I remembered @exbinary was interested in, but I might have misremembered

File: app/filters/contact_method_filter.rb
Comment (exbinary) [inline app/filters/contact_method_filter.rb:11]: I'm noticing all the filters use `parameters.keys`. Could we get the keys in `BrowseFilter` or `BaseFilter.new` and make just the keys available to individual filters?

Relatedly, the shape of `parameters` is not clear from looking at the code. Could be solved by looking at specs, or maybe a comment describing how params will come in?

File: app/filters/category_filter.rb
Comment (exbinary) [inline app/filters/category_filter.rb:7]: Serializing inside the filters doesn't feel quite right. Is it possible to have `options` return a relation as the value of `filters` and let the controller handle serialization?
```suggestion
    {
      name: "Categories",
      # Currently only filtering by top-level categories
      filters: Category.roots
    }
```

This might require introducing another blueprint at the level of a filter that then has an `association :filters`.

**Review by exbinary — APPROVED** [review]
Looks great, @h-m-m ! A few comments inline but overall really like how this is shaping.

File: app/filters/category_filter.rb
Comment (h-m-m) [reply app/filters/category_filter.rb]: The only with your suggestion is that the Ask/Offer distinction is one I'm doing manually:

https://github.com/rubyforgood/mutual-aid/blob/39a0d099aaa860edf396a000e86ee8bdb8397237/app/filters/contribution_type_filter.rb#L1-L7

File: app/filters/contact_method_filter.rb
Comment (h-m-m) [reply app/filters/contact_method_filter.rb:11]: > the shape of parameters is not clear from looking at the code

This is really the key point here. Thanks for surfacing this, because this point is helping me find the most confusing points.

I started writing some tests around this and got stuck because it was *hard*. I think I've introduced a small, modest mess in `BrowseFilter#contributions` and cleaning up the parameters structure here is dependent on cleaning that up, too.

There's a part of me that wants to get this PR out first and then focus on cleaning up that problem. I want to do that because this PR has taken a very long time and not directly for any code-specific reasons. This PR seems so big to me already, and I'd really like to focus on the refactor I'll need here as a separate thing from the other changes in this PR.

I'm open to suggestions. Let me know what you think

File: app/filters/contribution_type_filter.rb
Comment (maebeale) [reply app/filters/contribution_type_filter.rb:19]: yes. we'll need to add that scope to `CommunityResource` to return objects that are still published

File: app/controllers/contributions_controller.rb
Comment (h-m-m) [inline app/controllers/contributions_controller.rb:63]: TODO: rename `ALLOWED_PARAMS_FILTER`

File: app/models/browse_filter.rb
Comment (h-m-m) [inline app/models/browse_filter.rb:55]: TODO: this can probably be safely deleted

File: app/controllers/contributions_controller.rb
Comment (h-m-m) [inline app/controllers/contributions_controller.rb:18]: TODO: rename `filter_groups` -> `filter_groupings`

File: app/controllers/contributions_controller.rb
Comment (h-m-m) [reply app/controllers/contributions_controller.rb:63]: @exbinary suggested that calling it `ALLOWED_PARAMS` is probably sufficient

File: app/filters/category_filter.rb
Comment (exbinary) [reply app/filters/category_filter.rb]: Ah yes, i see the challenge. I had a go at extracting the serialization out in 8b8c8a94b5dd1e50f256e12abbd3071550ebed61, curious your thoughts?
In any case, i don't think this question should hold up this PR.

**Review by exbinary — COMMENTED** [review]
@h-m-m , i merged main, resolved conflicts and pushed to this branch.
Should be good to merge once you give it a quick look.

File: app/filters/contribution_type_filter.rb
Comment (h-m-m) [reply app/filters/contribution_type_filter.rb]: Note: discussed this with @exbinary in our regular meeting and added a commit per our discussion

**Comment by h-m-m** [comment]
Notes on what we're thinking of following up on as a subsequent refactoring:

- [ ] Removing `paramters.keys` references in the filters so filter classes can scan the full parameter list as they see fit ([comment](https://github.com/rubyforgood/mutual-aid/pull/546#discussion_r611060827)
- [ ] Move serialization concerns out of the filter classes, perhaps by breaking up `filter_grouping` in to one method that returns the filter grouping's name and another method that returns the filter options for that filter grouping ([comment](https://github.com/rubyforgood/mutual-aid/pull/546#discussion_r611322656))

**Pull request merged**
